app/src/main/java/io/apicurio/registry/storage/importing/v2/SqlDataUpgrader.java

```diff
@@ -0,0 +1,270 @@
+                        .owner(entity.createdBy).build();
+                storage.importArtifact(artifactEntity);
+            }
+
+            storage.importArtifactVersion(newEntity);
+            log.debug("Artifact version imported successfully: {}", entity);
+            globalIdMapping.put(oldGlobalId, entity.globalId);
+            var gav = new GAV(entity.groupId, entity.artifactId, entity.version);
+            gavDone.add(gav);
+
+            // Import comments that were waiting for this version
+            var commentsToImport = waitingForVersion.stream()
+                    .filter(comment -> comment.globalId == oldGlobalId).toList();
+            for (CommentEntity commentEntity : commentsToImport) {
+                importComment(commentEntity);
+            }
+            waitingForVersion.removeAll(commentsToImport);
+
+        } catch (VersionAlreadyExistsException ex) {
+            if (ex.getGlobalId() != null) {
+                log.warn("Duplicate globalId {} detected, skipping import of artifact version: {}",
+                        ex.getGlobalId(), entity);
+            } else {
+                log.warn("Failed to import artifact version {}: {}", entity, ex.getMessage());
+            }
+        } catch (Exception ex) {
+            log.warn("Failed to import artifact version {}: {}", entity, ex.getMessage());
+        }
+    }
+
+    @Override
+    public void importContent(ContentEntity entity) {
+        try {
+            List<ArtifactReferenceDto> references = SqlUtil
+                    .deserializeReferences(entity.serializedReferences);
+
+            TypedContent typedContent = TypedContent.create(ContentHandle.create(entity.contentBytes), null);
+            Map<String, TypedContent> resolvedReferences = storage.resolveReferences(references);
+            entity.artifactType = utils.determineArtifactType(typedContent, null, resolvedReferences);
+
+            // First we have to recalculate both the canonical hash and the contentHash
+            TypedContent canonicalContent = utils.canonicalizeContent(entity.artifactType, typedContent,
+                    resolvedReferences);
+
+            entity.canonicalHash = DigestUtils.sha256Hex(canonicalContent.getContent().bytes());
+            entity.contentHash = utils.getContentHash(typedContent, references);
+
+            // Then, based on the configuration, a new id is requested or the old one is used.
+            var oldContentId = entity.contentId;
+            if (!preserveContentId) {
+                entity.contentId = storage.nextContentId();
+            }
+
+            // Finally, using the information from the old content, a V3 content entity is created.
+            io.apicurio.registry.utils.impexp.v3.ContentEntity newEntity = io.apicurio.registry.utils.impexp.v3.ContentEntity
+                    .builder().contentType(entity.artifactType).contentHash(entity.contentHash)
```

Comment:
Content type and artifact type are not the same.  The artifact type is going to be OPENAPI or ASYNCAPI or AVRO I think.  The Content type needs to be `application/json` or `application/yaml` or `text/xml`.  We need logic here to figure out the content type from the content probably.  Or we could try to make a guess based on the artifact type (AVRO is `application/json`).  I think making a guess might work OK for a v2 upgrade, since we always converted YAML to JSON in v2.

Reply:
Ok, this is what I thought, but I would have expected some test to fail. I'll add some logic to guess the content type based on the type.

Reply:
You would get weird results from the "Get Version Content" operation, because it will set the HTTP response `Content-Type` to e.g. `AVRO`.  Which might be handled gracefully in many places.

The content type is mostly used inbound (e.g. parsing an OpenAPI document when applying rules).  We don't apply any rules during import though, so insane values for content type would work fine.